### ISSUE TITLE
Allow for visual_size offset in set_attach

### ISSUE THREAD

**Issue by EliasFleckenstein03** [minetest/minetest#10986]
## Problem
When attaching an object to another, the position, rotation and visual_size is inherited, but you can provide offsets for position and rotation. In most cases the inheritation of visual_size is unwanted; for example if you attach a player with a visual size of x=1 y=1 z=1 to a boat with a visual size of x=3 y=3 z=3, you don't want the player's size to be x=1*3=3 y=1*3=3 z=1*3=3. So, if you want to attach an object to another, you always need to change the child's visual_size to it's original visual size divided by the parent's visual size, and when detaching you have to revert it. This is annoying and slows down development.
## Solutions
Extend `ObjectRef:set_attach(parent, bone, position_offset, rotation_offset, forced_visible)` to `ObjectRef:set_attach(parent, bone, position_offset, rotation_offset, forced_visible, size_offset)`

## Alternatives
Make visual_size inheritance optional by extending `ObjectRef:set_attach(parent, bone, position_offset, rotation_offset)` to `ObjectRef:set_attach(parent, bone, position_offset, rotation_offset, forced_visible, inherit_size)`

**Comment by rubenwardy** [comment]
An alternative would be to deprecate `visual_size` and add two new things:

* `scale` as an object variable, like with position and rotation. For compat, it would also be used to set `visual_size` and vise-versa.
* another object property which only effects the current object

This would be closer to what other scene graph software does

**Comment by EliasFleckenstein03** [comment]
Could you "rank" all the suggestions based on estimated required effort / how likely it is that it will be merged? I'll start working on the best.

**Comment by v-rob** [comment]
It could be nice to actually transform objects using 4x4 transformation matrices.  That could probably solve the problems with `set/get_rotation` (although I admit I'm no expert on transformation matricies).

**Comment by EliasFleckenstein03** [comment]
I think position + rotation + scale is enough for almost all cases. A transformation matrix would be a bit of an overkill.

**Comment by rubenwardy** [comment]
> I think position + rotation + scale is enough for almost all cases. A transformation matrix would be a bit of an overkill.

You typically implement position rotation and scale using a transformation matrix, irrlicht will be doing this under the hood

**Comment by EliasFleckenstein03** [comment]
> You typically implement position rotation and scale using a transformation matrix, irrlicht will be doing this under the hood

Yes, I know that - what I meant was that it would be an overkill to let mod developers provide transformation matrices.

**Comment by v-rob** [comment]
The idea behind transformation matricies is that they can solve problems like gimbal locking with rotation, or making rotation around a specific axis easier, or choosing which order to apply rotation, translation, and scale, etc.  (If you've used `set/get_rotation` with the existing vector helpers, you'll understand why rotation is hard to work with currently.)  Certainly, you could also do other things, like skewing, but I don't consider those the useful features of transformation matricies.

However, using a transformation matrix isn't much harder in its basic form either.  For instance, without:
```lua
obj:set_rotation(vec)
obj:set_pos(vec)
obj:set_scale(vec)
```
With:
```lua
local m = matrix4.new()
m:set_rotation(vec)
m:set_translation(vec)
m:set_scale(vec)
obj:set_transformation_matrix(m)
```

**Comment by rubenwardy** [comment]
> Could you "rank" all the suggestions based on estimated required effort / how likely it is that it will be merged? I'll start working on the best suggestion afterwards.

What I said seems much easier than adding new things to attachment hierarchies. The problem would be backwards compatibility when mods use the new size object property, but there's similar issues with the other options

Also, exposing a transformation matrix API to Lua code isn't really relevant for this issue - but it does make sense to use them in the engine

**Comment by EliasFleckenstein03** [comment]
> However, using a transformation matrix isn't much harder in its basic form either. For instance, without:

The problem I see with this are collision/selection boxes. Since they are *axis-aligned*, rotation should not be applied to them (and I don't think we want them to scale with the visual size either). Therefore, we'd have to extract the translation from the transform matrix again when translating the boxes (which is in itself easy but kind of still unnecessary).
There are also tons of other cases where you need to access just the position of an object. Creating a transformation matrix server-side means merging variables together at a point where you still need to access them separately.
